File: client/src/main/java/com/flipkart/grayskull/models/GrayskullClientConfiguration.java
package com.flipkart.grayskull.models;

import lombok.Getter;

/**
 * Configuration properties for the Grayskull client.
 * <p>
 * This class holds all the necessary configuration parameters required to connect
 * to and interact with the Grayskull.
 * </p>
 */
@Getter
public final class GrayskullClientConfiguration {
    
    /**
     * The Grayskull server endpoint URL.
     * <p>
     * This should be the DNS of the Grayskull server (e.g., "https://grayskull.example.com").
     * The client will append appropriate API paths to this host.
     * </p>
     */
    private String host;
    
    /**
     * The connection timeout in milliseconds.
     * <p>
     * This is the maximum time the client will wait when establishing a connection
     * to the Grayskull server. If the connection cannot be established within this time,
     * a timeout exception will be thrown.
     * </p>
     * <p>
     * Default: 10000ms (10 seconds)
     * </p>
     */
    private int connectionTimeout = 10000;
    
    /**
     * The read timeout in milliseconds.
     * <p>
     * This is the maximum time the client will wait for data to be received from the
     * Grayskull server after a connection has been established. If no data is received
     * within this time, a timeout exception will be thrown.
     * </p>
     * <p>
     * Default: 30000ms (30 seconds)
     * </p>
     */
    private int readTimeout = 30000;
    
    /**
     * The maximum number of concurrent connections to maintain.
     * <p>
     * This controls the connection pool size for the HTTP client. More connections
     * allow for more concurrent requests but consume more resources.
     * </p>
     * <p>
     * Default: 10
     * </p>
     */
    private int maxConnections = 10;
    
    /**
     * The maximum number of retry attempts for failed requests.
     * <p>
     * When a request to the Grayskull server fails due to transient errors
     * (e.g., network issues, server temporarily unavailable), the client will
     * retry the request up to this many times before giving up.
     * </p>
     * <p>
     * Default: 3
     * </p>
     */
    private int maxRetries = 3;

    /**
     * The minimum delay in milliseconds between retry attempts.
     * <p>
     * When a request fails and needs to be retried, the client will wait at least
     * this amount of time before attempting the next retry. This helps prevent
     * overwhelming the server with rapid retry attempts.
     * </p>
     * <p>
     * Default: 100ms
     * </p>
     */
    private int minRetryDelay = 100;

    /**
     * Whether to enable metrics collection.
     * <p>
     * When enabled, the client will expose metrics about API calls (success/failure counts,
     * response times, etc.) via JMX MBeans/Micrometer for monitoring and observability.
     * </p>
     * <p>
     * Default: true
     * </p>
     */
    private boolean enableMetrics = true;

    /**
     * Sets the Grayskull server endpoint URL.
     *
     * @param host the server URL (must not be null or empty)
     * @throws IllegalArgumentException if host is null or empty
     */
    public void setHost(String host) {
        if (host == null || host.trim().isEmpty()) {
            throw new IllegalArgumentException("Host cannot be null or empty");
        }
        // Remove trailing slash if present
        this.host = host.endsWith("/") ? host.substring(0, host.length() - 1) : host;
    }

    /**
     * Sets the connection timeout in milliseconds.
     *
     * @param connectionTimeout the connection timeout in milliseconds (must be positive)
     * @throws IllegalArgumentException if connectionTimeout is not positive
     */
    public void setConnectionTimeout(int connectionTimeout) {
        if (connectionTimeout <= 0) {
            throw new IllegalArgumentException("Connection timeout must be positive, got: " + connectionTimeout);
        }
        this.connectionTimeout = connectionTimeout;
    }

    /**
     * Sets the read timeout in milliseconds.
     *
     * @param readTimeout the read timeout in milliseconds (must be positive)
     * @throws IllegalArgumentException if readTimeout is not positive
     */
    public void setReadTimeout(int readTimeout) {
        if (readTimeout <= 0) {
            throw new IllegalArgumentException("Read timeout must be positive, got: " + readTimeout);
        }
        this.readTimeout = readTimeout;
    }

    /**
     * Sets the maximum number of concurrent connections.
     *
     * @param maxConnections the maximum number of connections (must be positive)
     * @throws IllegalArgumentException if maxConnections is not positive
     */
    public void setMaxConnections(int maxConnections) {
        if (maxConnections <= 0) {
            throw new IllegalArgumentException("Max connections must be positive, got: " + maxConnections);
        }
        this.maxConnections = maxConnections;
    }

    /**
     * Sets the maximum number of retry attempts.
     *
     * @param maxRetries the maximum number of retry attempts (must be between 1 and 10)
     * @throws IllegalArgumentException if maxRetries is less than 1 or greater than 10
     */
    public void setMaxRetries(int maxRetries) {
        if (maxRetries < 1) {
            throw new IllegalArgumentException("Max retries cannot be less than 1, got: " + maxRetries);
        }
        if (maxRetries > 10) {
            throw new IllegalArgumentException("Max retries cannot be greater than 10, got: " + maxRetries);
        }
        this.maxRetries = maxRetries;
    }

    /**
     * Sets the minimum delay between retry attempts in milliseconds.
     *
     * @param minRetryDelay the minimum retry delay in milliseconds (must be at least 50ms)
     * @throws IllegalArgumentException if minRetryDelay is less than 50ms
     */
    public void setMinRetryDelay(int minRetryDelay) {
        if (minRetryDelay < 50) {
            throw new IllegalArgumentException("Min retry delay must be at least 50ms, got: " + minRetryDelay);
        }
        this.minRetryDelay = minRetryDelay;
    }

    /**
     * Sets whether to enable JMX metrics collection.
     *
     * @param enableMetrics true to enable metrics, false to disable
     */
    public void setEnableMetrics(boolean enableMetrics) {
        this.enableMetrics = enableMetrics;
    }
}

File: clients/go/client_impl/models/grayskullClientConfiguration.go
package models

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

// GrayskullClientConfiguration holds all configuration parameters required to connect
// to and interact with the Grayskull service.
type GrayskullClientConfiguration struct {
	// Host is the Grayskull server endpoint URL (e.g., "https://grayskull.example.com").
	// The client will append appropriate API paths to this host.
	Host string

	// ConnectionTimeout is the maximum duration for establishing a connection to the Grayskull server.
	// Default: 10 seconds
	ConnectionTimeout time.Duration

	// ReadTimeout is the maximum duration for reading the entire response.
	// Default: 30 seconds
	ReadTimeout time.Duration

	// MaxConnections controls the maximum number of concurrent connections.
	// Default: 10
	MaxConnections int

	// MaxRetries is the maximum number of retry attempts for failed requests.
	// Must be between 1 and 10.
	// Default: 3
	MaxRetries int

	// MinRetryDelay is the minimum delay between retry attempts.
	// Must be at least 50ms.
	// Default: 100ms
	MinRetryDelay time.Duration

	// MetricsEnabled controls whether metrics collection is enabled.
	// Default: true
	MetricsEnabled bool
}

// NewGrayskullClientConfiguration creates a new configuration with default values.
func NewGrayskullClientConfiguration() *GrayskullClientConfiguration {
	return &GrayskullClientConfiguration{
		ConnectionTimeout: 10 * time.Second,
		ReadTimeout:       30 * time.Second,
		MaxConnections:    10,
		MaxRetries:        3,
		MinRetryDelay:     100 * time.Millisecond,
		MetricsEnabled:    true,
	}
}

// SetHost sets the Grayskull server endpoint URL.
// Returns an error if host is empty.
func (c *GrayskullClientConfiguration) SetHost(host string) error {
	if strings.TrimSpace(host) == "" {
		return errors.New("host cannot be empty")
	}
	// Remove trailing slash if present
	c.Host = strings.TrimSuffix(host, "/")
	return nil
}

// SetConnectionTimeout sets the connection timeout.
// Returns an error if the timeout is not positive.
func (c *GrayskullClientConfiguration) SetConnectionTimeout(timeout time.Duration) error {
	if timeout <= 0 {
		return errors.New("connection timeout must be positive")
	}
	c.ConnectionTimeout = timeout
	return nil
}

// SetReadTimeout sets the read timeout.
// Returns an error if the timeout is not positive.
func (c *GrayskullClientConfiguration) SetReadTimeout(timeout time.Duration) error {
	if timeout <= 0 {
		return errors.New("read timeout must be positive")
	}
	c.ReadTimeout = timeout
	return nil
}

// SetMaxConnections sets the maximum number of concurrent connections.
// Returns an error if maxConnections is not positive.
func (c *GrayskullClientConfiguration) SetMaxConnections(maxConnections int) error {
	if maxConnections <= 0 {
		return errors.New("max connections must be positive")
	}
	c.MaxConnections = maxConnections
	return nil
}

// SetMaxRetries sets the maximum number of retry attempts.
// Returns an error if maxRetries is not between 1 and 10.
func (c *GrayskullClientConfiguration) SetMaxRetries(maxRetries int) error {
	if maxRetries < 1 || maxRetries > 10 {
		return fmt.Errorf("max retries must be between 1 and 10, got %d", maxRetries)
	}
	c.MaxRetries = maxRetries
	return nil
}

// SetMinRetryDelay sets the minimum delay between retry attempts.
// Returns an error if the delay is less than 50ms.
func (c *GrayskullClientConfiguration) SetMinRetryDelay(delay time.Duration) error {
	if delay < 50*time.Millisecond {
		return errors.New("min retry delay must be at least 50ms")
	}
	c.MinRetryDelay = delay
	return nil
}

// SetMetricsEnabled enables or disables metrics collection.
func (c *GrayskullClientConfiguration) SetMetricsEnabled(enabled bool) {
	c.MetricsEnabled = enabled
}
